File: app/src/main/java/com/kotlinblog/actors/view/MainActivity.java
package com.kotlinblog.actors.view;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import com.kotlinblog.actors.R;

import timber.log.Timber;

public class MainActivity extends AppCompatActivity {

    private MainViewModel mViewModel;
    private MainAdapter mAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mViewModel = ViewModelProviders.of(this).get(MainViewModel.class);
        mAdapter = new MainAdapter(mViewModel, this);

        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        LinearLayoutManager lm = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(lm);
        recyclerView.setAdapter(mAdapter);

        DividerItemDecoration divider = new DividerItemDecoration(
                recyclerView.getContext(),
                lm.getOrientation()
        );

        recyclerView.addItemDecoration(divider);

        if (savedInstanceState == null) {
            mViewModel.fetchActors();
        }

        observeActors();
    }

    private void observeActors() {
        mViewModel.getActors().observe(this,
                actorsList -> {
                    mAdapter.notifyDataSetChanged();
                    Timber.d("Actors changed...");
                });

        mViewModel.getConnectionError().observe(this,
                error -> Toast.makeText(MainActivity.this,
                        "Connection error: " + error,
                        Toast.LENGTH_LONG)
                        .show());
    }
}

File: app/src/main/java/com/kotlinblog/actors/view/MainActivity.kt
package com.kotlinblog.actors.view

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast
import com.kotlinblog.actors.R
import kotlinx.android.synthetic.main.activity_main.*
import timber.log.Timber

class MainActivity : AppCompatActivity() {

    private lateinit var mViewModel: MainViewModel
    private lateinit var mAdapter: MainAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mViewModel = ViewModelProviders.of(this).get(MainViewModel::class.java)

        mAdapter = MainAdapter(mViewModel, applicationContext)
        val lm = LinearLayoutManager(this)

        recyclerView.apply {
            layoutManager = lm
            adapter = mAdapter
            val divider = DividerItemDecoration(
                    recyclerView.context,
                    lm.orientation)
            addItemDecoration(divider)
        }

        if (savedInstanceState == null) {
            mViewModel.fetchActors()
        }
        observeActors()
    }

    private fun observeActors() {
        mViewModel.actors.observe(this, Observer {
            mAdapter.notifyDataSetChanged()
            Timber.d("Actors changed...")
        })

        mViewModel.connectionError.observe(this,
                Observer {
                    Toast.makeText(this@MainActivity,
                            "Connection error: $it",
                            Toast.LENGTH_LONG)
                            .show()
                })
    }
}
